### DISCUSSION TITLE
simple solution: modify value of relationship field within afterRead hook

### DISCUSSION THREAD

**Discussion by unonweb** [payloadcms/payload#1684] [Q&A] [answered]
Hi, 

as there are some threads about the _relationship & afterRead hook_ matter I'd like to share my findings:

Related Threads:
* [Possibility to only have certain fields included in relationship](https://github.com/payloadcms/payload/discussions/1042)
* [Return of afterRead hook for relationship field](https://github.com/payloadcms/payload/discussions/1354)

The relationship field:
* stores `relationTo: 'collection-slug'` and `value: 'collection-id'`
* is populated _after_ the afterRead hook

The afterRead hook...
* is called _both_ by the admin UI _and_ server side functions
* is called _before_ relationships are populated
* may return a modified result

**If we want to modify the value of the relationship field when it's queried** (which means in the afterRead hook), we face the following issue:
We need to include the original relationship data when the admin UI reads the field (otherwise it breaks).
But _if_ we include `relationTo` and `value` properties, the relationship field is going to be populated automatically.

The solutions is:
1. Set `maxDepth: 0` which practically disables the automatic population of the field.
2. Populate the field manually whithin a afterRead hook

```js
const relationshipField = {
	type: 'relationship',
	name: 'internal',
	label: 'Document to link to',
	relationTo: ['pages'],
	required: true,
	maxDepth: 0, // <-- important
	hooks: {
		afterRead: [
			(args) => extractFieldsFromRelationship(args, ['url'])
		]
	}
}

async function extractFieldsFromRelationship(args, topLevelFields = []) {
	// afterRead

	if (typeof args.value === 'undefined') return

	const refCollection = args.value.relationTo
	const refID = args.value.value

	const doc = await payload.findByID({
		collection: refCollection,
		id: refID,
		depth: 1,
	})
	
	const extract = {}

	topLevelFields.map(field => {
		if (doc[field]) extract[field] = doc[field]
	})

	// include this - otherwise the admin UI will not work properly
	extract.relationTo = refCollection 
	extract.value = refID

	//return doc // This is the default way. The whole doc is included.
	return extract
}
```

**Comment by jacobsfletch** [comment, marked as answer]
@unonweb this is great!! The only thing that jumps out at me are that it seems like you should _not_ have to set `maxDepth: 0` — because this requires you to make that extra query from within your hook. The entire document is already present in the args, so couldn't you just strip out the irrelevant fields without querying your relationship again?

That first discussion you linked has some pretty good information, for others who stumble by. Namely, doing this will break GraphQL, and that this is really where GraphQL shines.

Thank you for this thorough example!

**Comment by christian-reichart** [comment]
Thanks for that tip!
Also related, solves that really elegantly: https://github.com/payloadcms/payload/pull/1288

I think using REST has a lot of benefits over GraphQL for certain cases, but it's currently handicapped by the relationships because of this select issue, also when using an internal link inside the Rich Text for example.
